**Name:** Tamayo, Ray Lan A.<br>
**Course and Section:** CPE019-CPE32S9 <br>
**Date Submission:** 03/20/2024 <br>
**Instructor:** Engr. Roman Richard <br>

In this assignment, you are task to build a multilayer perceptron model. The following are the requirements:



# **Choose any dataset**  
* **Diabetes Dataset**

# **Explain the problem you are trying to solve**

* This dataset is originally from the National Institute of Diabetes and Digestive and Kidney Diseases. The objective is to predict based on diagnostic measurements whether a patient has diabetes.

# **Create your own model**
**Importing Libraries**

In [53]:
# Importing Libraries

import pandas as pd
import numpy as np
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier


**Loading Data**

In [58]:
# Load the Data

dbts = pd.read_csv('/content/diabetes.csv')
print(dbts.head())

unique_outcomes = dbts['Outcome'].unique()
print(unique_outcomes)

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  
[1 0]


* <p align="justify"> The dataset consists of various health-related measurements for individuals, with the 'Outcome' column indicating whether each person has diabetes or not.

In [57]:
# Show the row and columns

shape_of_dbts = dbts.shape
print(shape_of_dbts)

(768, 9)


* The dataset contains records for 768 individuals, each described by 9 different health-related attributes.

In [59]:
# Summary statistics function

description_of_dbts = dbts.describe()
print(description_of_dbts)

       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  

* <p align="justify"> The descriptive statistics reveal that the dataset includes information about 768 individuals' health attributes, such as pregnancies, glucose levels, and blood pressure, with various measures of central tendency and dispersion.

In [60]:
# Separating features and target

X = dbts.drop(columns='Outcome')
y = dbts['Outcome']
print(X.head(2))
print(y.head(2))

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   

   DiabetesPedigreeFunction  Age  
0                     0.627   50  
1                     0.351   31  
0    1
1    0
Name: Outcome, dtype: int64


* <p align="justify"> The dataset is split into two parts: 'X' containing various health-related measurements for individuals, and 'y' representing whether each person has diabetes or not.

**Spliting the Data**

In [61]:
# Splitting data for modeling

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)

Shape of X_train: (537, 8)
Shape of X_test: (231, 8)


* <p align="justify"> The dataset is divided into training and testing sets, with 537 samples in the training set and 231 samples in the testing set, each containing 8 features.

**Training the Model**

In [62]:
# MLP Classifier configuration

mlp = MLPClassifier()
mlp.hidden_layer_sizes = (3, 2)
mlp.max_iter = 500
mlp.activation = 'relu'

* <p align="justify"> The MLP classifier is configured to have two hidden layers with 3 and 2 neurons respectively, allowing up to 500 iterations for training, and using the ReLU activation function.

In [63]:
# Training MLP model

mlp = MLPClassifier().fit(X_train, y_train)

* <p align="justify"> The MLP classifier is being trained on the training data to learn patterns between the input features (X_train) and their corresponding target labels (y_train).

**Testing**

In [64]:
# Making test predictions

pred = mlp.predict(X_test)
print(pred)

[0 0 0 0 1 0 1 1 0 0 0 1 0 0 0 0 1 1 0 0 0 0 1 1 0 0 0 1 0 0 0 0 0 0 0 1 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 1 0 1 1 1 1 0 1 0 1 0 0 1 0
 1 0 0 0 1 1 1 0 0 1 1 1 0 0 0 0 0 0 0 1 1 0 0 0 1 1 0 0 0 0 0 1 1 0 0 0 0
 0 1 1 0 0 0 0 0 0 0 1 0 1 0 1 0 0 1 0 0 1 0 0 0 0 0 0 1 1 0 0 0 1 1 1 0 1
 1 0 0 0 0 1 1 0 0 1 0 0 0 1 0 0 0 0 1 0 1 0 0 1 1 0 1 0 0 0 0 1 0 0 1 0 0
 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 1 0 0 1 0 1 1 0 0 0 1 0
 0 0 0 1 1 1 0 1 1]


* <p align="justify"> The model has made predictions on the test data, indicating whether each instance is classified as having diabetes (1) or not having diabetes (0) based on the learned patterns from the training data.

# **Evaluate the accuracy of your model**

In [67]:
# Computing confusion matrix

conf_matrix = confusion_matrix(y_test, pred)
print(conf_matrix)

[[120  25]
 [ 40  46]]


* <p align="justify"> The confusion matrix shows that 139 instances were correctly predicted as not having diabetes, 7 instances were correctly predicted as having diabetes, 83 instances were falsely predicted as not having diabetes, and 2 instances were falsely predicted as having diabetes.

In [68]:
# Model evaluation summary

report = classification_report(y_test, pred)
print(report)

              precision    recall  f1-score   support

           0       0.75      0.83      0.79       145
           1       0.65      0.53      0.59        86

    accuracy                           0.72       231
   macro avg       0.70      0.68      0.69       231
weighted avg       0.71      0.72      0.71       231



* <p align="justify"> Model's performance summary, indicating precision, recall, F1-score, and accuracy for predicting diabetes presence or absence, with higher values desired for better performance.

# **Conclusion**

* <p align="justify">
In this assignment, I developed a multilayer perceptron (MLP) model to tackle a specific problem. I chose the problem of predicting diabetes based on health-related measurements. By training the MLP model on a dataset containing various health metrics, I aimed to create a predictive tool that could help identify individuals at risk of diabetes. Evaluating the model's accuracy provided insights into its effectiveness in correctly classifying individuals as either having or not having diabetes, which is crucial for its practical applicability in healthcare scenarios.
